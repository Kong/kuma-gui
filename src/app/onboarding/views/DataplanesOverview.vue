<template>
  <OnboardingPage>
    <template #header>
      <OnboardingHeading
        :title="title"
        :description="description"
      />
    </template>

    <template #content>
      <div
        v-if="!tableData.data.length"
        class="justify-center flex my-4"
      >
        <LoadingBox />
      </div>

      <div v-else>
        <div class="flex justify-center mt-10 mb-16 pb-16">
          <div class="w-full sm:w-3/5 p-4">
            <p class="font-bold mb-4">
              Found {{ tableData.data.length }} DPPs:
            </p>

            <KTable
              class="onboarding-dataplane-table"
              :fetcher="() => tableData"
              :headers="tableHeaders"
              disable-pagination
              is-small
            >
              <template #status="{ rowValue }">
                <div
                  class="entity-status"
                  :class="{
                    'is-offline': rowValue.toLowerCase() === 'offline' || rowValue === false,
                    'is-online': rowValue.toLowerCase() === 'online',
                    'is-degraded': rowValue.toLowerCase() === 'partially degraded',
                    'is-not-available': rowValue.toLowerCase() === 'not available',
                  }"
                >
                  <span>{{ rowValue }}</span>
                </div>
              </template>
            </KTable>
          </div>
        </div>
      </div>
    </template>

    <template #navigation>
      <OnboardingNavigation
        next-step="onboarding-completed"
        previous-step="onboarding-add-services-code"
        :should-allow-next="tableData.data.length > 0"
      />
    </template>
  </OnboardingPage>
</template>

<script>
import { KTable } from '@kong/kongponents'

import { PRODUCT_NAME, OFFLINE } from '@/constants'
import { getItemStatusFromInsight } from '@/utilities/dataplane'
import { kumaApi } from '@/api/kumaApi'
import LoadingBox from '@/app/common/LoadingBox.vue'
import OnboardingNavigation from '../components/OnboardingNavigation.vue'
import OnboardingHeading from '../components/OnboardingHeading.vue'
import OnboardingPage from '../components/OnboardingPage.vue'

export default {
  name: 'DataplanesOverview',
  components: {
    OnboardingNavigation,
    OnboardingHeading,
    OnboardingPage,
    LoadingBox,
    KTable,
  },
  metaInfo() {
    return {
      title: this.title,
    }
  },
  data() {
    return {
      productName: PRODUCT_NAME,
      tableHeaders: [
        { label: 'Mesh', key: 'mesh' },
        { label: 'Name', key: 'name' },
        { label: 'Status', key: 'status' },
      ],
      tableData: {
        total: 0,
        data: [],
      },
      timeout: null,
    }
  },

  computed: {
    title() {
      if (this.tableData.data.length) {
        return 'Success'
      }

      return 'Waiting for DPPs'
    },
    description() {
      if (this.tableData.data.length) {
        return 'The following data plane proxies (DPPs) are connected to the control plane:'
      }

      return null
    },
  },
  created() {
    this.getAllDataplanes()
  },
  beforeUnmount() {
    clearTimeout(this.timeout)
  },
  methods: {
    async getAllDataplanes() {
      let shouldRefetch = false
      const result = []

      try {
        const { items } = await kumaApi.getAllDataplanes({ size: 10 })

        if (Array.isArray(items)) {
          for (let i = 0; i < items.length; i++) {
            const { name, mesh } = items[i]

            const { status } = await kumaApi.getDataplaneOverviewFromMesh({ mesh, name }).then((response) =>
              getItemStatusFromInsight(response.dataplaneInsight),
            )

            if (status === OFFLINE) {
              shouldRefetch = true
            }

            result.push({
              status,
              name,
              mesh,
            })
          }
        }
      } catch (e) {
        console.error(e)
      }

      this.tableData.data = result
      this.tableData.total = this.tableData.data.length

      if (shouldRefetch) {
        this.timeout = setTimeout(() => {
          this.getAllDataplanes()
        }, 1000)
      }
    },
  },
}
</script>

<style lang="scss">
.onboarding-dataplane-table tbody tr {
  // hack to do not allow shadow overflow on top/bottom
  box-shadow: -3px 0 0 -1px var(--KTableBorder, var(--steel-200, #dae3f2)) !important;
}
</style>
