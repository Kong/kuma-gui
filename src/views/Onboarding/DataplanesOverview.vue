<template>
  <OnboardingPage>
    <template #header>
      <OnboardingHeading
        :title="title"
        :description="description"
      />
    </template>
    <template #content>
      <div
        v-if="!tableData.data.length"
        class="justify-center flex my-4"
      >
        <LoadingBox />
      </div>
      <div v-else>
        <div class="flex justify-center mt-10 mb-16 pb-16">
          <div class="w-full sm:w-3/5 p-4">
            <p class="font-bold mb-4">
              Found {{ tableData.data.length }} DPPs:
            </p>
            <KTable
              class="onboarding-dataplane-table"
              :fetcher="() => tableData"
              :headers="tableHeaders"
              disable-pagination
              is-small
            >
              <template #status="{ rowValue }">
                <div
                  class="entity-status"
                  :class="{
                    'is-offline': rowValue.toLowerCase() === 'offline' || rowValue === false,
                    'is-online': rowValue.toLowerCase() === 'online',
                    'is-degraded': rowValue.toLowerCase() === 'partially degraded',
                    'is-not-available': rowValue.toLowerCase() === 'not available',
                  }"
                >
                  <span>{{ rowValue }}</span>
                </div>
              </template>
            </KTable>
          </div>
        </div>
      </div>
    </template>

    <template #navigation>
      <OnboardingNavigation
        next-step="onboarding-completed"
        previous-step="onboarding-add-services-code"
        :should-allow-next="tableData.data.length > 0"
      />
    </template>
  </OnboardingPage>
</template>

<script>
import { PRODUCT_NAME, OFFLINE } from '@/consts'
import { getItemStatusFromInsight } from '@/dataplane'
import Kuma from '@/services/kuma'
import LoadingBox from '@/components/LoadingBox.vue'
import OnboardingNavigation from '@/views/Onboarding/components/OnboardingNavigation.vue'
import OnboardingHeading from '@/views/Onboarding/components/OnboardingHeading.vue'
import OnboardingPage from '@/views/Onboarding/components/OnboardingPage.vue'

export default {
  name: 'DataplanesOverview',
  components: {
    OnboardingNavigation,
    OnboardingHeading,
    OnboardingPage,
    LoadingBox,
  },
  metaInfo() {
    return {
      title: this.title,
    }
  },
  data() {
    return {
      productName: PRODUCT_NAME,
      tableHeaders: [
        { label: 'Mesh', key: 'mesh' },
        { label: 'Name', key: 'name' },
        { label: 'Status', key: 'status' },
      ],
      tableData: {
        total: 0,
        data: [],
      },
      timeout: null,
    }
  },

  computed: {
    title() {
      if (this.tableData.data.length) {
        return 'Success'
      }

      return 'Waiting for DPPs'
    },
    description() {
      if (this.tableData.data.length) {
        return 'The following data plane proxies (DPPs) are connected to the control plane:'
      }

      return null
    },
  },
  created() {
    this.getAllDataplanes()
  },
  beforeUnmount() {
    clearTimeout(this.timeout)
  },
  methods: {
    async getAllDataplanes() {
      let shouldRefetch = false
      const result = []

      try {
        const dataplanes = await Kuma.getAllDataplanes({ size: 10 })
        const items = dataplanes.items

        for (let i = 0; i < items.length; i++) {
          const { name, mesh } = items[i]

          const { status } = await Kuma.getDataplaneOverviewFromMesh({ mesh, name }).then((response) =>
            getItemStatusFromInsight(response.dataplaneInsight),
          )

          if (status === OFFLINE) {
            shouldRefetch = true
          }

          result.push({
            status,
            name,
            mesh,
          })
        }
      } catch (e) {
        console.error(e)
      }

      this.tableData.data = result
      this.tableData.total = this.tableData.data.length

      if (shouldRefetch) {
        this.timeout = setTimeout(() => {
          this.getAllDataplanes()
        }, 1000)
      }
    },
  },
}
</script>

<style lang="scss">
.onboarding-dataplane-table tbody tr {
  // hack to do not allow shadow overflow on top/bottom
  box-shadow: -3px 0 0 -1px var(--KTableBorder, var(--steel-200, #dae3f2)) !important;
}
</style>
