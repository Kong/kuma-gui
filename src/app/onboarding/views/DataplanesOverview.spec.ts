import { describe, expect, test } from '@jest/globals'
import { flushPromises, mount } from '@vue/test-utils'

import DataplanesOverview from './DataplanesOverview.vue'

function renderComponent() {
  return mount(DataplanesOverview)
}

describe('DataplanesOverview.vue', () => {
  test('renders snapshot', async () => {
    const wrapper = renderComponent()

    expect(wrapper.find('[data-testid="loading"]').exists()).toBe(true)
    expect(wrapper.find('h1').html()).toContain('Waiting for DPPs')
    expect(wrapper.find('[data-testid="onboarding-next-button"]').attributes('disabled')).toBe('true')

    await flushPromises()
    expect(wrapper.html()).toContain('dataplane-test-456')

    expect(wrapper.element).toMatchSnapshot()
  })
})
