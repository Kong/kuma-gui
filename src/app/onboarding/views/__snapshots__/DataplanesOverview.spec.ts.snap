// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DataplanesOverview.vue renders snapshot 1`] = `
<div
  class="route-view"
>
  <div
    aria-atomic="true"
    aria-live="assertive"
    class="route-view-title visually-hidden"
  >
    Navigated to Data plane overview
  </div>
  
  <!--v-if-->
  <main
    class="app-main-content"
  >
    
    <!--v-if-->
    <section
      class=""
    >
      <!--v-if-->
      
      <div>
        <div
          class="onboarding-container"
        >
          <div
            class="onboarding-container__header"
          >
            
            <div
              class="onboarding-heading"
            >
              <h1
                class="onboarding-title"
              >
                
                <p>
                  Success
                </p>
                
              </h1>
              <div
                class="onboarding-description"
              >
                
                <p>
                  The following data plane proxies (DPPs) are connected to the control plane:
                </p>
                
              </div>
            </div>
            
          </div>
          <div
            class="onboarding-container__content"
          >
            <div
              class="onboarding-container__inner-content"
            >
              
              <div>
                <p
                  class="mb-4"
                >
                  <b>
                    Found 8 DPPs:
                  </b>
                </p>
                <div
                  class="k-table-container mb-4"
                >
                  <!---->
                  <section
                    class="k-table-wrapper"
                  >
                    <table
                      class="k-table has-hover"
                      data-tableid="aaaabbbb-cccc-dddd-eeee-ffffffffffff"
                    >
                      <thead
                        class=""
                      >
                        <tr
                          class=""
                        >
                          
                          <th
                            class="k-table-headers"
                          >
                            <span
                              class="k-table-headers-container"
                            >
                              
                              <span
                                class=""
                              >
                                Mesh
                              </span>
                              
                              <!---->
                            </span>
                          </th>
                          <th
                            class="k-table-headers"
                          >
                            <span
                              class="k-table-headers-container"
                            >
                              
                              <span
                                class=""
                              >
                                Name
                              </span>
                              
                              <!---->
                            </span>
                          </th>
                          <th
                            class="k-table-headers"
                          >
                            <span
                              class="k-table-headers-container"
                            >
                              
                              <span
                                class=""
                              >
                                Status
                              </span>
                              
                              <!---->
                            </span>
                          </th>
                          
                        </tr>
                      </thead>
                      <tbody>
                        
                        <tr>
                          
                          <td>
                            
                            matrix-0
                            
                          </td>
                          <td>
                            
                            dataplane-test-456
                            
                          </td>
                          <td>
                            
                            <div
                              class="k-badge k-badge-danger k-badge-rounded status"
                              data-testid="status-badge"
                              tabindex="0"
                            >
                              <div
                                class="k-badge-text"
                              >
                                <div
                                  class="k-badge-text"
                                >
                                  
                                  offline
                                  
                                </div>
                              </div>
                              <!---->
                            </div>
                            
                          </td>
                          
                        </tr>
                        <tr>
                          
                          <td>
                            
                            capacitor-1
                            
                          </td>
                          <td>
                            
                            driver-1
                            
                          </td>
                          <td>
                            
                            <div
                              class="k-badge k-badge-danger k-badge-rounded status"
                              data-testid="status-badge"
                              tabindex="0"
                            >
                              <div
                                class="k-badge-text"
                              >
                                <div
                                  class="k-badge-text"
                                >
                                  
                                  offline
                                  
                                </div>
                              </div>
                              <!---->
                            </div>
                            
                          </td>
                          
                        </tr>
                        <tr>
                          
                          <td>
                            
                            panel-2
                            
                          </td>
                          <td>
                            
                            card-2
                            
                          </td>
                          <td>
                            
                            <div
                              class="k-badge k-badge-danger k-badge-rounded status"
                              data-testid="status-badge"
                              tabindex="0"
                            >
                              <div
                                class="k-badge-text"
                              >
                                <div
                                  class="k-badge-text"
                                >
                                  
                                  offline
                                  
                                </div>
                              </div>
                              <!---->
                            </div>
                            
                          </td>
                          
                        </tr>
                        <tr>
                          
                          <td>
                            
                            matrix-3
                            
                          </td>
                          <td>
                            
                            panel-3
                            
                          </td>
                          <td>
                            
                            <div
                              class="k-badge k-badge-danger k-badge-rounded status"
                              data-testid="status-badge"
                              tabindex="0"
                            >
                              <div
                                class="k-badge-text"
                              >
                                <div
                                  class="k-badge-text"
                                >
                                  
                                  offline
                                  
                                </div>
                              </div>
                              <!---->
                            </div>
                            
                          </td>
                          
                        </tr>
                        <tr>
                          
                          <td>
                            
                            program-4
                            
                          </td>
                          <td>
                            
                            bandwidth-4
                            
                          </td>
                          <td>
                            
                            <div
                              class="k-badge k-badge-danger k-badge-rounded status"
                              data-testid="status-badge"
                              tabindex="0"
                            >
                              <div
                                class="k-badge-text"
                              >
                                <div
                                  class="k-badge-text"
                                >
                                  
                                  offline
                                  
                                </div>
                              </div>
                              <!---->
                            </div>
                            
                          </td>
                          
                        </tr>
                        <tr>
                          
                          <td>
                            
                            interface-5
                            
                          </td>
                          <td>
                            
                            microchip-5
                            
                          </td>
                          <td>
                            
                            <div
                              class="k-badge k-badge-danger k-badge-rounded status"
                              data-testid="status-badge"
                              tabindex="0"
                            >
                              <div
                                class="k-badge-text"
                              >
                                <div
                                  class="k-badge-text"
                                >
                                  
                                  offline
                                  
                                </div>
                              </div>
                              <!---->
                            </div>
                            
                          </td>
                          
                        </tr>
                        <tr>
                          
                          <td>
                            
                            interface-6
                            
                          </td>
                          <td>
                            
                            array-6
                            
                          </td>
                          <td>
                            
                            <div
                              class="k-badge k-badge-danger k-badge-rounded status"
                              data-testid="status-badge"
                              tabindex="0"
                            >
                              <div
                                class="k-badge-text"
                              >
                                <div
                                  class="k-badge-text"
                                >
                                  
                                  offline
                                  
                                </div>
                              </div>
                              <!---->
                            </div>
                            
                          </td>
                          
                        </tr>
                        <tr>
                          
                          <td>
                            
                            feed-7
                            
                          </td>
                          <td>
                            
                            interface-7
                            
                          </td>
                          <td>
                            
                            <div
                              class="k-badge k-badge-danger k-badge-rounded status"
                              data-testid="status-badge"
                              tabindex="0"
                            >
                              <div
                                class="k-badge-text"
                              >
                                <div
                                  class="k-badge-text"
                                >
                                  
                                  offline
                                  
                                </div>
                              </div>
                              <!---->
                            </div>
                            
                          </td>
                          
                        </tr>
                        
                      </tbody>
                    </table>
                    <!---->
                  </section>
                </div>
              </div>
              
            </div>
          </div>
          <div
            class="mt-4"
          >
            
            <div
              class="onboarding-actions"
            >
              <a
                class="k-button medium rounded secondary"
                data-testid="onboarding-previous-button"
                href="/gui/onboarding/add-services-code"
                type="button"
              >
                
                <!---->
                
                
                 Back 
                
                <!---->
              </a>
              <div
                class="button-list"
              >
                <a
                  aria-current="page"
                  class="router-link-active router-link-exact-active k-button medium rounded outline"
                  data-testid="onboarding-skip-button"
                  href="/gui/"
                  type="button"
                >
                  
                  <!---->
                  
                  
                   Skip setup 
                  
                  <!---->
                </a>
                <a
                  class="k-button medium rounded primary"
                  data-testid="onboarding-next-button"
                  href="/gui/onboarding/completed"
                  type="button"
                >
                  
                  <!---->
                  
                  
                  Next
                  
                  <!---->
                </a>
              </div>
            </div>
            
          </div>
        </div>
        <div
          class="background-image"
        />
      </div>
      
    </section>
    
  </main>
  
</div>
`;
